So this is 'fun' - we dont seem to have `std::filesystem` yet in our version of C++ (C++17 IIRC is where that came in), so i tried using the `experimental::filesystem` code and while it _does_ build the relevant file, it dies in linkage with
```make
  CXXLD    src/libvyatta-cfg.la
/usr/bin/ld: src/cstore/unionfs/.libs/cstore-unionfs.o: in function `std::experimental::filesystem::v1::copy(std::experimental::filesystem::v1::__cxx11::path const&, std::experimental::filesystem::v1::__cxx11::path const&)':
/usr/include/c++/10/experimental/bits/fs_ops.h:62: undefined reference to `std::experimental::filesystem::v1::copy(std::experimental::filesystem::v1::__cxx11::path const&, std::experimental::filesystem::v1::__cxx11::path const&, std::experimental::filesystem::v1::copy_options)'
/usr/bin/ld: /usr/include/c++/10/experimental/bits/fs_ops.h:62: undefined reference to `std::experimental::filesystem::v1::copy(std::experimental::filesystem::v1::__cxx11::path const&, std::experimental::filesystem::v1::__cxx11::path const&, std::experimental::filesystem::v1::copy_options)'
/usr/bin/ld: src/cstore/unionfs/.libs/cstore-unionfs.o: in function `std::experimental::filesystem::v1::__cxx11::path::path<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::experimental::filesystem::v1::__cxx11::path>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/include/c++/10/experimental/bits/fs_path.h:230: undefined reference to `std::experimental::filesystem::v1::__cxx11::path::_M_split_cmpts()'
/usr/bin/ld: src/cstore/unionfs/.libs/cstore-unionfs.o: in function `std::experimental::filesystem::v1::__cxx11::path::path<char const*, std::experimental::filesystem::v1::__cxx11::path>(char const* const&)':
/usr/include/c++/10/experimental/bits/fs_path.h:230: undefined reference to `std::experimental::filesystem::v1::__cxx11::path::_M_split_cmpts()'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:844: src/libvyatta-cfg.la] Error 1
make[3]: Leaving directory '/vyos/packages/vyatta-cfg'
make[2]: *** [Makefile:1414: all-recursive] Error 1
make[2]: Leaving directory '/vyos/packages/vyatta-cfg'
make[1]: *** [Makefile:642: all] Error 2
make[1]: Leaving directory '/vyos/packages/vyatta-cfg'
dh_auto_build: error: make -j1 returned exit code 2
make: *** [debian/rules:24: build] Error 255
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
```

--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,9 @@
+vyatta-cfg (0.102.0+vyos2+current5) unstable; urgency=medium
+
+  * Provide internal stream_copy fall-through for Boost's copy_file
+
+ -- Boris Lukashev <rageltman@sempervictus.com>  Sat, 13 Aug 2022 19:00:00 -0500
+
 vyatta-cfg (0.102.0+vyos2+current4) unstable; urgency=medium
 
   * Adjust for replacement of Quagga with FRR.

--- a/src/cstore/unionfs/cstore-unionfs.cpp
+++ b/src/cstore/unionfs/cstore-unionfs.cpp
@@ -167,6 +167,14 @@ _unescape_path_name(const string& path)
   return npath;
 }
 
+// Fall-through for Boost's filesystem::copy_file "complexity"
+void stream_file( const char* srce_file, const char* dest_file )
+{
+    std::ifstream srce( srce_file, std::ios::binary ) ;
+    std::ofstream dest( dest_file, std::ios::binary ) ;
+    dest << srce.rdbuf() ;
+}
+
 vector<int> getActiveCommits()
 {
   string process_name = "vbash";
@@ -321,7 +329,6 @@ UnionfsCstore::UnionfsCstore(const string& sid, string& env)
 UnionfsCstore::~UnionfsCstore()
 {
 }
-
 
 ////// public virtual functions declared in base class
 bool
@@ -750,7 +757,12 @@ UnionfsCstore::sync_dir(const FsPath& src, const FsPath& dst,
       try {
         if (path_is_regular(s)) {
           // it's file
-          b_fs::copy_file(s.path_cstr(), d.path_cstr());
+          try {
+            b_fs::copy_file(s.path_cstr(), d.path_cstr());
+          } catch (const boost::filesystem::filesystem_error& e) {
+            output_internal("syncdir failed due to %s in copy_file. Falling back to internal stream_file\n", e.what());
+            stream_file(s.path_cstr(), d.path_cstr());
+          }
         } else {
           // dir
           recursive_copy_dir(s, d, true);
@@ -1592,7 +1604,12 @@ UnionfsCstore::recursive_copy_dir(const FsPath& src, const FsPath& dst,
           }
         }
       }
-      b_fs::copy_file(di->path(), nname);
+      try {
+        b_fs::copy_file(di->path(), nname);
+      } catch (const b_fs::filesystem_error& e) {
+        output_internal("recursive_copy_dir failed due to %s in copy_file. Falling back to internal stream_file\n", e.what());
+        stream_file(di->path().string().c_str(), nname.c_str());
+      }
     }
   }
 }

